### PR TITLE
Fix for #976 to Ignore Unused Stock Variables

### PR DESCRIPTION
**Pull request by c0rp3n**
# Solution

This change looks to ignore stock variables that are not from the main source file.

# Testing

Added two new compile tests for stock variables and arrays.

### CODE REVIEW

**Comment by dvander** [comment]
This error is intentional. The warning should only elide on includes.

**Comment by A1mDev** [comment]
#976

**Comment by c0rp3n** [comment]
Updated to elide the warning on includes.
I could also look to have the same behaviour for #972 but they would likely want to share code.

File: compiler/semantics.cpp [hunk 1]
```diff
@@ -2307,6 +2307,18 @@ bool Semantics::CheckExprStmt(ExprStmt* stmt) {
...
 
+bool Semantics::IsIncluded(Decl* sym) {
+    const auto fileno = cc().sources()->GetSourceFileIndex(sym->pos());
+    if (fileno >= cc().sources()->opened_files().size()) { return false; }
+
+    return ! cc().sources()->opened_files()[fileno]->is_main_file();
+}
+
+template <typename DeclType>
+bool Semantics::IsIncludedStock(DeclType* sym) {
```
Comment (dvander) [inline compiler/semantics.cpp:2318]: This should be VarDeclBase, no template needed.

File: compiler/semantics.cpp [hunk 2]
```diff
@@ -2307,6 +2307,18 @@ bool Semantics::CheckExprStmt(ExprStmt* stmt) {
     return true;
 }
 
+bool Semantics::IsIncluded(Decl* sym) {
+    const auto fileno = cc().sources()->GetSourceFileIndex(sym->pos());
```
Comment (dvander) [inline compiler/semantics.cpp:2311]: cc() should be cc_, here and elsewhere.

File: compiler/semantics.cpp [hunk 3]
```diff
@@ -2307,6 +2307,18 @@ bool Semantics::CheckExprStmt(ExprStmt* stmt) {
...
+    if (fileno >= cc().sources()->opened_files().size()) { return false; }
+
+    return ! cc().sources()->opened_files()[fileno]->is_main_file();
```
Comment (dvander) [inline compiler/semantics.cpp:2314]: nit: "!cc_" (no space)

File: compiler/semantics.cpp [hunk 4]
```diff
@@ -2307,6 +2307,18 @@ bool Semantics::CheckExprStmt(ExprStmt* stmt) {
     return true;
 }
 
+bool Semantics::IsIncluded(Decl* sym) {
+    const auto fileno = cc().sources()->GetSourceFileIndex(sym->pos());
+    if (fileno >= cc().sources()->opened_files().size()) { return false; }
```
Comment (dvander) [inline compiler/semantics.cpp:2312]: nit: no single-line conditionals. should be:

    if (fileno >= cc().sources()->opened_files().size()
        return false;

Though, I think this condition is unnecessary. We should never get a source index for a file not in the open list.

**Comment by A1mDev** [comment]
The problem still exists with global variables, could you fix your PR?

**Comment by dvander** [comment]
I'll fix up and land this.

**Pull request closed without merging**